### PR TITLE
[REEF-1721] Close the job submission handler when REEF client closes.

### PR DESCRIPTION
**Pull request by motus**
Also add some extra logging in the `REEF.close()` call.

JIRA: [REEF-1721](https://issues.apache.org/jira/browse/REEF-1721)

Closes #

### CODE REVIEW

**Comment by yunseong** [comment]
LGTM. I'll test and merge it. Thanks!

**Comment by tcNickolas** [comment]
@motus Do you have some logs to show running threads which need to be closed by closing job submission handler? The change looks good to me but I'm not familiar with Java threading in REEF enough to predict the behavior change caused by this.

**Comment by motus** [comment]
I will try to reproduce it. It seems to be a sporadic event in certain
runtimes (e.g. in local mode; for YARN the submission handler does not
require any cleanup).

On Wed, Jan 18, 2017 at 5:14 PM, Mariia Mykhailova <notifications@github.com

**Comment by yunseong** [comment]
@tcNickolas Thanks for pointing it out! @motus I'll wait a few hours before merging it :)

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/client/REEFImplementation.java [hunk 1]
```diff
@@ -89,8 +90,21 @@
...
+
+    LOG.log(Level.FINE, "Close REEF: shutdown jobs");
     this.runningJobs.closeAllJobs();
+
+    LOG.log(Level.FINE, "Close REEF: shutdown client");
     this.clientWireUp.close();
+
+    LOG.log(Level.FINE, "Close REEF: shutdown job submitter");
+    try {
+      this.jobSubmissionHandler.close();
```
Comment (jwang98052) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/client/REEFImplementation.java]: Do you see any error or improper shut down if we don't call jobSubmissionHandler.close()?

Comment (yunseong) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/client/REEFImplementation.java]: @motus gentle ping :) Have you succeded to reproduce the sporadic event?

@tcNickolas @jwang98052 In my opinion, we don't have to skip `jobSubmissionHandler.close()`, though we haven't seen any problemetic case without calling it. Is there anything I'm missing?

**Comment by motus** [comment]
@yunseong @jwang98052 Here's a log from HelloREEF client with local runtime, when we do *not* do that final `.close()` call:

```
2017-01-23 15:31:31,949 FINE reef.util.ThreadLogger.logThreads main | Threads running at the end of HelloREEF:
        TG main THREAD main :: RUNNABLE, Alive, NOT interrupted, NOT daemon
                java.lang.Thread.dumpThreads(Native Method)
                java.lang.Thread.getAllStackTraces(Thread.java:1603)
                org.apache.reef.util.ThreadLogger.getFormattedThreadList(ThreadLogger.java:79)
                org.apache.reef.util.ThreadLogger.logThreads(ThreadLogger.java:62)
                org.apache.reef.util.ThreadLogger.logThreads(ThreadLogger.java:45)
                org.apache.reef.examples.hello.HelloREEF.main(HelloREEF.java:70)
        TG main THREAD org.apache.reef.wake.remote.impl.RemoteSenderStage:pool-06:thread-001 :: WAITING, Alive, NOT interrupted, NOT daemon
                java.lang.Object.wait(Native Method)
                java.lang.Object.wait(Object.java:502)
                io.netty.util.concurrent.DefaultPromise.awaitUninterruptibly(DefaultPromise.java:286)
                io.netty.channel.DefaultChannelPromise.awaitUninterruptibly(DefaultChannelPromise.java:135)
                io.netty.channel.DefaultChannelPromise.awaitUninterruptibly(DefaultChannelPromise.java:28)
                io.netty.util.concurrent.DefaultPromise.syncUninterruptibly(DefaultPromise.java:225)
                io.netty.channel.DefaultChannelPromise.syncUninterruptibly(DefaultChannelPromise.java:123)
                io.netty.channel.DefaultChannelPromise.syncUninterruptibly(DefaultChannelPromise.java:28)
                org.apache.reef.wake.remote.transport.netty.NettyMessagingTransport.open(NettyMessagingTransport.java:300)
                org.apache.reef.wake.remote.impl.ConnectCallable.call(RemoteSenderEventHandler.java:139)
                org.apache.reef.wake.remote.impl.ConnectCallable.call(RemoteSenderEventHandler.java:125)
                java.util.concurrent.FutureTask.run(FutureTask.java:266)
                java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
                java.util.concurrent.FutureTask.run(FutureTask.java:266)
                java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
                java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
                java.lang.Thread.run(Thread.java:745)
```
Normal behavior is to have only the `main` thread running at the end of the client process.

**Comment by yunseong** [comment]
@motus Thanks for the result, and additional PR for reproducing this issue. 

When I ran HelloREEF with this PR and #1239 merged, I got the following result:
```
2017-01-25 09:49:04,215 FINE reef.util.ThreadLogger.logThreads main | Threads running at the end of HelloREEF:
	TG main THREAD globalEventExecutor-1-6 :: TIMED_WAITING, Alive, NOT interrupted, NOT daemon
		sun.misc.Unsafe.park(Native Method)
		java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
		java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2078)
		java.util.concurrent.LinkedBlockingQueue.poll(LinkedBlockingQueue.java:467)
		io.netty.util.concurrent.GlobalEventExecutor.takeTask(GlobalEventExecutor.java:90)
		io.netty.util.concurrent.GlobalEventExecutor$TaskRunner.run(GlobalEventExecutor.java:334)
		io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:137)
		java.lang.Thread.run(Thread.java:745)
	TG main THREAD main :: RUNNABLE, Alive, NOT interrupted, NOT daemon
		java.lang.Thread.dumpThreads(Native Method)
		java.lang.Thread.getAllStackTraces(Thread.java:1603)
		org.apache.reef.util.ThreadLogger.getFormattedThreadList(ThreadLogger.java:79)
		org.apache.reef.util.ThreadLogger.logThreads(ThreadLogger.java:62)
		org.apache.reef.util.ThreadLogger.logThreads(ThreadLogger.java:45)
		org.apache.reef.examples.hello.HelloREEF.main(HelloREEF.java:70)
	TG main THREAD pool-1-thread-1 :: WAITING, Alive, NOT interrupted, NOT daemon
		java.lang.Object.wait(Native Method)
		java.lang.Object.wait(Object.java:502)
		java.lang.UNIXProcess.waitFor(UNIXProcess.java:396)
		org.apache.reef.runtime.local.process.RunnableProcess.run(RunnableProcess.java:213)
		java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
		java.util.concurrent.FutureTask.run(FutureTask.java:266)
		java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
		java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
		java.lang.Thread.run(Thread.java:745)
```
I have two questions:
1. The first thread appears intermittently. Maybe the thread is from the Netty, which shuts down gracefully?
2. Is it a normal behavior to have the last entry (`TG main THREAD pool-1-thread-1 :: WAITING, Alive, NOT interrupted, NOT daemon`)? I'd like to confirm that your result does not show one.

**Comment by motus** [comment]
@yunseong yes, `globalEventExecutor-1-6` is an internal Wake/Netty thread; it's OK to exit when this thread still exists and is in `TIMED_WAIT` state.

Another thread `pool-1-thread-1` I don't like. I made an effort in my previous pull requests to give more meaningful names to our threads so we can tell where exactly we've started them. I will check why this thread still has an old name and fix it in subsequent PRs.

Also, @yunseong could you please tell me if your HelloREEF client had exited normally or it just hung after dumping the thread info? A non-daemon thread in `WAITING` mode would prevent Java process from exiting, so if your process exits fine, then everything is fine, and the thread showing up in the logs is just a timing issue that we can safely ignore. If your process blocks on exit because of this thread - we have an issue. Please check that [REEF-1715](https://issues.apache.org/jira/browse/REEF-1715) is **not** merged in to make sure that the client is not waiting on the Driver - it looks like that `pool-1-thread-1` thread is actually waiting for the Driver process to finish.

**Comment by jwang98052** [comment]
LGTM. I think we should close jobSubmissionHandler in any case. 
@yunseong is it Ok for the merge?

**Comment by yunseong** [comment]
@motus The client always finished successfully. I'll merge this PR  as it works as expected.

**Pull request closed without merging**